x-pack/plugins/endpoint/public/applications/endpoint/store/alerts/selectors.ts

```diff
@@ -32,48 +41,113 @@ export const isOnAlertPage = (state: AlertListState): boolean => {
-  if (state.location) {
-    // Removes the `?` from the beginning of query string if it exists
-    const query = qs.parse(state.location.search.slice(1));
-    return {
-      ...(query.page_size ? { page_size: query.page_size } : {}),
-      ...(query.page_index ? { page_index: query.page_index } : {}),
-    };
-  } else {
-    return {};
+export const uiQueryParams: (
+  state: AlertListState
+) => Immutable<AlertingIndexUIQueryParams> = createSelector(
+  (state: AlertListState) => state.location,
+  (location: AlertListState['location']) => {
+    const data: AlertingIndexUIQueryParams = {};
+    if (location) {
+      // Removes the `?` from the beginning of query string if it exists
+      const query = qs.parse(location.search.slice(1));
+
+      /**
+       * Build an AlertingIndexUIQueryParams object with keys from the query.
+       * If more than one value exists for a key, use the last.
+       */
+      const keys: Array<keyof AlertingIndexUIQueryParams> = [
+        'page_size',
+        'page_index',
+        'selected_alert',
+      ];
+      for (const key of keys) {
+        const value = query[key];
+        if (typeof value === 'string') {
+          data[key] = value;
+        } else if (Array.isArray(value)) {
+          data[key] = value[value.length - 1];
+        }
+      }
+    }
+    return data;
   }
-};
+);
+
+/**
+ * query params to use when requesting alert data.
+ */
+export const apiQueryParams: (
+  state: AlertListState
+) => Immutable<AlertsAPIQueryParams> = createSelector(
+  uiQueryParams,
+  ({ page_size, page_index }) => ({
+    page_size,
+    page_index,
+  })
+);
 
 /**
  * Returns a function that takes in a new page size and returns a new query param string
  */
 export const urlFromNewPageSizeParam: (
   state: AlertListState
-) => (newPageSize: number) => string = state => {
+) => (newPageSize: number) => string = createSelector(uiQueryParams, paramData => {
   return newPageSize => {
-    const urlPaginationData = paginationDataFromUrl(state);
-    urlPaginationData.page_size = newPageSize.toString();
+    const queryParams: AlertingIndexUIQueryParams = { ...paramData };
+    queryParams.page_size = newPageSize.toString();
 
-    // Only set the url back to page zero if the user has changed the page index already
-    if (urlPaginationData.page_index !== undefined) {
-      urlPaginationData.page_index = '0';
+    /**
+     * Reset the page index when changing page size.
+     */
+    if (queryParams.page_index !== undefined) {
+      delete queryParams.page_index;
     }
-    return '?' + qs.stringify(urlPaginationData);
+    return '?' + qs.stringify(queryParams);
   };
-};
+});
 
 /**
  * Returns a function that takes in a new page index and returns a new query param string
  */
 export const urlFromNewPageIndexParam: (
   state: AlertListState
-) => (newPageIndex: number) => string = state => {
+) => (newPageIndex: number) => string = createSelector(uiQueryParams, paramData => {
```

Comment:
@alexk307 I missed your earlier comment, so replying here.
Selectors have a common interface, they take the state of the app (what is returned by the top level reducer) and return whatever they want. The return value should be a primitive, or treated as immutable as its used to invalidate views and things.

These selectors want to take a second param. In this case, the page number for the link they want. But since selectors always take exactly one argument, we use a thunk instead.

Helpers that take a selector, e.g. `useSelector`, `createStructuredSelector`, `createSelector`, etc, all still work fine, since we are still a standard (odd) selector. Its not a pattern i like to use too often, but we've got a whole bunch here (and in Resolver.)